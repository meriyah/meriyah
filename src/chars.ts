/**
 * A list of character constants with much more human-readable names.
 */
export const enum Chars {

    Null = 0x00,

    Backspace      = 0x08,
    Tab            = 0x09,
    LineFeed       = 0x0A,
    VerticalTab    = 0x0B,
    FormFeed       = 0x0C,
    CarriageReturn = 0x0D,

    Space       = 0x20,
    Exclamation = 0x21,
    DoubleQuote = 0x22,
    Hash        = 0x23,
    Dollar      = 0x24,
    Percent     = 0x25,
    Ampersand   = 0x26,
    SingleQuote = 0x27,
    LeftParen   = 0x28,
    RightParen  = 0x29,
    Asterisk    = 0x2A,
    Plus        = 0x2B,
    Comma       = 0x2C,
    Hyphen      = 0x2D,
    Period      = 0x2E,
    Slash       = 0x2F,

    Zero         = 0x30,
    One          = 0x31,
    Two          = 0x32,
    Three        = 0x33,
    Four         = 0x34,
    Five         = 0x35,
    Six          = 0x36,
    Seven        = 0x37,
    Eight        = 0x38,
    Nine         = 0x39,
    Colon        = 0x3A,
    Semicolon    = 0x3B,
    LessThan     = 0x3C,
    EqualSign    = 0x3D,
    GreaterThan  = 0x3E,
    QuestionMark = 0x3F,

    UpperA = 0x41,
    UpperB = 0x42,
    UpperC = 0x43,
    UpperD = 0x44,
    UpperE = 0x45,
    UpperF = 0x46,
    UpperG = 0x47,
    UpperH = 0x48,
    UpperI = 0x49,
    UpperJ = 0x4A,
    UpperK = 0x4B,
    UpperL = 0x4C,
    UpperM = 0x4D,
    UpperN = 0x4E,
    UpperO = 0x4F,

    UpperP       = 0x50,
    UpperQ       = 0x51,
    UpperR       = 0x52,
    UpperS       = 0x53,
    UpperT       = 0x54,
    UpperU       = 0x55,
    UpperV       = 0x56,
    UpperW       = 0x57,
    UpperX       = 0x58,
    UpperY       = 0x59,
    UpperZ       = 0x5A,
    LeftBracket  = 0x5B,
    Backslash    = 0x5C,
    RightBracket = 0x5D,
    Caret        = 0x5E,
    Underscore   = 0x5F,

    Backtick = 0x60,
    LowerA   = 0x61,
    LowerB   = 0x62,
    LowerC   = 0x63,
    LowerD   = 0x64,
    LowerE   = 0x65,
    LowerF   = 0x66,
    LowerG   = 0x67,
    LowerH   = 0x68,
    LowerI   = 0x69,
    LowerJ   = 0x6A,
    LowerK   = 0x6B,
    LowerL   = 0x6C,
    LowerM   = 0x6D,
    LowerN   = 0x6E,
    LowerO   = 0x6F,

    LowerP      = 0x70,
    LowerQ      = 0x71,
    LowerR      = 0x72,
    LowerS      = 0x73,
    LowerT      = 0x74,
    LowerU      = 0x75,
    LowerV      = 0x76,
    LowerW      = 0x77,
    LowerX      = 0x78,
    LowerY      = 0x79,
    LowerZ      = 0x7A,
    LeftBrace   = 0x7B,
    VerticalBar = 0x7C,
    RightBrace  = 0x7D,
    Tilde       = 0x7E,
    Delete      = 0x7F,

    NextLine         = 0x85,
    NonBreakingSpace = 0xA0,

    Ogham = 0x1680,

    EnQuad             = 0x2000,
    EmQuad             = 0x2001,
    EnSpace            = 0x2002,
    EmSpace            = 0x2003,
    ThreePerEmSpace    = 0x2004,
    FourPerEmSpace     = 0x2005,
    SixPerEmSpace      = 0x2006,
    FigureSpace        = 0x2007,
    PunctuationSpace   = 0x2008,
    ThinSpace          = 0x2009,
    HairSpace          = 0x200A,
    ZeroWidthSpace     = 0x200B,
    ZeroWidthJoiner    = 0x200C,
    ZeroWidthNonJoiner = 0x200D,

    LineSeparator      = 0x2028,
    ParagraphSeparator = 0x2029,

    NarrowNoBreakSpace = 0x202F,
    MathematicalSpace  = 0x205F,
    IdeographicSpace   = 0x3000,

    ZeroWidthNoBreakSpace = 0xFEFF,

    ByteOrderMark = 0xFFEF,

    NonBMPMax = 0x10FFFF,

    LeadSurrogateMin = 0xD800,
    LeadSurrogateMax = 0xDBFF,
    TrailSurrogateMin = 0xDC00,
    TrailSurrogateMax = 0xDFFF,
    UTF16Max = 0xFFFF,
    NonBMPMin = 0x10000,
}
