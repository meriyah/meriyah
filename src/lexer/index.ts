export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
