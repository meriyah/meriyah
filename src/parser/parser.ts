import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
