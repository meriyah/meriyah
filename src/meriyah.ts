import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
