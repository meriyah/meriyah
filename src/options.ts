import type * as ESTree from './estree';
import { pushComment, pushToken } from './parser/parser';
import { type Token } from './token';

/**
 * Function calls when semicolon inserted.
 */
type OnInsertedSemicolon = (pos: number) => any;

/**
 * Token process function.
 */
export type OnToken = (token: string, start: number, end: number, loc: ESTree.SourceLocation) => any;

/**
 * Comment process function.
 */
export type OnComment = (
  type: ESTree.CommentType,
  value: string,
  start: number,
  end: number,
  loc: ESTree.SourceLocation,
) => any;

/**
 * The parser options.
 */
export interface Options {
  // Allow module code
  module?: boolean;
  // Enable stage 3 support (ESNext)
  next?: boolean;
  // Enable start and end offsets to each node
  ranges?: boolean;
  // Enable web compatibility
  webcompat?: boolean;
  // Enable line/column location information to each node
  loc?: boolean;
  // Attach raw property to each literal and identifier node
  raw?: boolean;
  // Allow return in the global scope
  globalReturn?: boolean;
  // Enable implied strict mode
  impliedStrict?: boolean;
  // Enable non-standard parenthesized expression node
  preserveParens?: boolean;
  // Enable lexical binding and scope tracking
  lexical?: boolean;
  // Adds a source attribute in every nodeâ€™s loc object when the locations option is `true`
  source?: string;
  // Enable React JSX parsing
  jsx?: boolean;
  // Allows comment extraction. Accepts either a callback function or an array
  onComment?: ESTree.Comment[] | OnComment;
  // Allows detection of automatic semicolon insertion. Accepts a callback function that will be passed the character offset where the semicolon was inserted
  onInsertedSemicolon?: OnInsertedSemicolon;
  // Allows token extraction. Accepts either a callback function or an array
  onToken?: Token[] | OnToken;
}

export type NormalizedOptions = Omit<Options, 'onComment' | 'onToken'> & {
  onComment?: OnComment;
  onToken?: OnToken;
};

export function normalizeOptions(rawOptions: Options): NormalizedOptions {
  const options = { ...rawOptions } as NormalizedOptions;

  // Accepts either a callback function to be invoked or an array to collect comments (as the node is constructed)
  if (options.onComment) {
    options.onComment = Array.isArray(options.onComment) ? pushComment(options.onComment, options) : options.onComment;
  }

  // Accepts either a callback function to be invoked or an array to collect tokens
  if (options.onToken) {
    options.onToken = Array.isArray(options.onToken) ? pushToken(options.onToken, options) : options.onToken;
  }

  return options;
}
