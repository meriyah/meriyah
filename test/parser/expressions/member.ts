import * as t from 'node:assert/strict';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { fail, pass } from '../../test-utils';
describe('Expressions - Member', () => {
  fail('Expressions - Member (fail)', ['abc.123', 'a.[b].c().d.toString()', 'abc.£', 'abc???.£']);

  for (const arg of [
    'let f = () => { import("foo"); };',
    'foo["bar"];',
    'foo.bar;',
    'foo.bar.foo;',
    'foo.bar["foo"];',
    'foo["foo"]["bar"];',
    'foo[test()][bar()];',
    '0..toString();',
    '0.5.toString();',
    '1.0.toString();',
    '1.000.toString();',
    'x.void',
    String.raw`x.voi\u0064`,
    'x.protected',
    String.raw`x.prot\u0065cted`,
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { next: true, sourceType: 'module' });
      });
    });
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { next: true });
      });
    });
  }

  pass('Expressions - Member (pass)', [
    'abc.package',
    { code: 'abc.package', options: { sourceType: 'module' } },
    { code: 'x[a, b]', options: { ranges: true } },
    { code: '(2[x,x],x)>x', options: { ranges: true } },
    'foo.bar',
    { code: '(a[b]||(c[d]=e))', options: { ranges: true } },
    'a&&(b=c)&&(d=e)',
    { code: 'a.$._.B0', options: { ranges: true } },
    'a.if',
    { code: 'a().b', options: { ranges: true } },
    'x.y / z',
    'a[b, c]',
    { code: 'a[b]||(c[d]=e)', options: { ranges: true } },
    'a&&(b=c)',
  ]);
});
