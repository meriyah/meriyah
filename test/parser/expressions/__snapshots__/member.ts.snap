// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expressions - Member > Expressions - Member (fail) > a.[b].c().d.toString() 1`] = `
"SyntaxError [1:2-1:3]: Dot property must be an identifier
> 1 | a.[b].c().d.toString()
    |   ^ Dot property must be an identifier"
`;

exports[`Expressions - Member > Expressions - Member (fail) > abc.123 1`] = `
"SyntaxError [1:3-1:7]: Unexpected token: 'number'
> 1 | abc.123
    |    ^^^^ Unexpected token: 'number'"
`;

exports[`Expressions - Member > Expressions - Member (fail) > abc.£ 1`] = `
"SyntaxError [1:4-1:4]: Illegal character '£'
> 1 | abc.£
    |     ^ Illegal character '£'"
`;

exports[`Expressions - Member > Expressions - Member (fail) > abc???.£ 1`] = `
"SyntaxError [1:5-1:7]: Unexpected token: '?.'
> 1 | abc???.£
    |      ^^ Unexpected token: '?.'"
`;

exports[`Expressions - Member > Expressions - Member (pass) > (2[x,x],x)>x 1`] = `
{
  "body": [
    {
      "end": 12,
      "expression": {
        "end": 12,
        "left": {
          "end": 9,
          "expressions": [
            {
              "computed": true,
              "end": 7,
              "object": {
                "end": 2,
                "range": [
                  1,
                  2,
                ],
                "start": 1,
                "type": "Literal",
                "value": 2,
              },
              "optional": false,
              "property": {
                "end": 6,
                "expressions": [
                  {
                    "end": 4,
                    "name": "x",
                    "range": [
                      3,
                      4,
                    ],
                    "start": 3,
                    "type": "Identifier",
                  },
                  {
                    "end": 6,
                    "name": "x",
                    "range": [
                      5,
                      6,
                    ],
                    "start": 5,
                    "type": "Identifier",
                  },
                ],
                "range": [
                  3,
                  6,
                ],
                "start": 3,
                "type": "SequenceExpression",
              },
              "range": [
                1,
                7,
              ],
              "start": 1,
              "type": "MemberExpression",
            },
            {
              "end": 9,
              "name": "x",
              "range": [
                8,
                9,
              ],
              "start": 8,
              "type": "Identifier",
            },
          ],
          "range": [
            1,
            9,
          ],
          "start": 1,
          "type": "SequenceExpression",
        },
        "operator": ">",
        "range": [
          0,
          12,
        ],
        "right": {
          "end": 12,
          "name": "x",
          "range": [
            11,
            12,
          ],
          "start": 11,
          "type": "Identifier",
        },
        "start": 0,
        "type": "BinaryExpression",
      },
      "range": [
        0,
        12,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 12,
  "range": [
    0,
    12,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > (a[b]||(c[d]=e)) 1`] = `
{
  "body": [
    {
      "end": 16,
      "expression": {
        "end": 15,
        "left": {
          "computed": true,
          "end": 5,
          "object": {
            "end": 2,
            "name": "a",
            "range": [
              1,
              2,
            ],
            "start": 1,
            "type": "Identifier",
          },
          "optional": false,
          "property": {
            "end": 4,
            "name": "b",
            "range": [
              3,
              4,
            ],
            "start": 3,
            "type": "Identifier",
          },
          "range": [
            1,
            5,
          ],
          "start": 1,
          "type": "MemberExpression",
        },
        "operator": "||",
        "range": [
          1,
          15,
        ],
        "right": {
          "end": 14,
          "left": {
            "computed": true,
            "end": 12,
            "object": {
              "end": 9,
              "name": "c",
              "range": [
                8,
                9,
              ],
              "start": 8,
              "type": "Identifier",
            },
            "optional": false,
            "property": {
              "end": 11,
              "name": "d",
              "range": [
                10,
                11,
              ],
              "start": 10,
              "type": "Identifier",
            },
            "range": [
              8,
              12,
            ],
            "start": 8,
            "type": "MemberExpression",
          },
          "operator": "=",
          "range": [
            8,
            14,
          ],
          "right": {
            "end": 14,
            "name": "e",
            "range": [
              13,
              14,
            ],
            "start": 13,
            "type": "Identifier",
          },
          "start": 8,
          "type": "AssignmentExpression",
        },
        "start": 1,
        "type": "LogicalExpression",
      },
      "range": [
        0,
        16,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 16,
  "range": [
    0,
    16,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > a&&(b=c) 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "a",
          "type": "Identifier",
        },
        "operator": "&&",
        "right": {
          "left": {
            "name": "b",
            "type": "Identifier",
          },
          "operator": "=",
          "right": {
            "name": "c",
            "type": "Identifier",
          },
          "type": "AssignmentExpression",
        },
        "type": "LogicalExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > a&&(b=c)&&(d=e) 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "left": {
            "name": "a",
            "type": "Identifier",
          },
          "operator": "&&",
          "right": {
            "left": {
              "name": "b",
              "type": "Identifier",
            },
            "operator": "=",
            "right": {
              "name": "c",
              "type": "Identifier",
            },
            "type": "AssignmentExpression",
          },
          "type": "LogicalExpression",
        },
        "operator": "&&",
        "right": {
          "left": {
            "name": "d",
            "type": "Identifier",
          },
          "operator": "=",
          "right": {
            "name": "e",
            "type": "Identifier",
          },
          "type": "AssignmentExpression",
        },
        "type": "LogicalExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > a().b 1`] = `
{
  "body": [
    {
      "end": 5,
      "expression": {
        "computed": false,
        "end": 5,
        "object": {
          "arguments": [],
          "callee": {
            "end": 1,
            "name": "a",
            "range": [
              0,
              1,
            ],
            "start": 0,
            "type": "Identifier",
          },
          "end": 3,
          "optional": false,
          "range": [
            0,
            3,
          ],
          "start": 0,
          "type": "CallExpression",
        },
        "optional": false,
        "property": {
          "end": 5,
          "name": "b",
          "range": [
            4,
            5,
          ],
          "start": 4,
          "type": "Identifier",
        },
        "range": [
          0,
          5,
        ],
        "start": 0,
        "type": "MemberExpression",
      },
      "range": [
        0,
        5,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 5,
  "range": [
    0,
    5,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > a.$._.B0 1`] = `
{
  "body": [
    {
      "end": 8,
      "expression": {
        "computed": false,
        "end": 8,
        "object": {
          "computed": false,
          "end": 5,
          "object": {
            "computed": false,
            "end": 3,
            "object": {
              "end": 1,
              "name": "a",
              "range": [
                0,
                1,
              ],
              "start": 0,
              "type": "Identifier",
            },
            "optional": false,
            "property": {
              "end": 3,
              "name": "$",
              "range": [
                2,
                3,
              ],
              "start": 2,
              "type": "Identifier",
            },
            "range": [
              0,
              3,
            ],
            "start": 0,
            "type": "MemberExpression",
          },
          "optional": false,
          "property": {
            "end": 5,
            "name": "_",
            "range": [
              4,
              5,
            ],
            "start": 4,
            "type": "Identifier",
          },
          "range": [
            0,
            5,
          ],
          "start": 0,
          "type": "MemberExpression",
        },
        "optional": false,
        "property": {
          "end": 8,
          "name": "B0",
          "range": [
            6,
            8,
          ],
          "start": 6,
          "type": "Identifier",
        },
        "range": [
          0,
          8,
        ],
        "start": 0,
        "type": "MemberExpression",
      },
      "range": [
        0,
        8,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 8,
  "range": [
    0,
    8,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > a.if 1`] = `
{
  "body": [
    {
      "expression": {
        "computed": false,
        "object": {
          "name": "a",
          "type": "Identifier",
        },
        "optional": false,
        "property": {
          "name": "if",
          "type": "Identifier",
        },
        "type": "MemberExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > a[b, c] 1`] = `
{
  "body": [
    {
      "expression": {
        "computed": true,
        "object": {
          "name": "a",
          "type": "Identifier",
        },
        "optional": false,
        "property": {
          "expressions": [
            {
              "name": "b",
              "type": "Identifier",
            },
            {
              "name": "c",
              "type": "Identifier",
            },
          ],
          "type": "SequenceExpression",
        },
        "type": "MemberExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > a[b]||(c[d]=e) 1`] = `
{
  "body": [
    {
      "end": 14,
      "expression": {
        "end": 14,
        "left": {
          "computed": true,
          "end": 4,
          "object": {
            "end": 1,
            "name": "a",
            "range": [
              0,
              1,
            ],
            "start": 0,
            "type": "Identifier",
          },
          "optional": false,
          "property": {
            "end": 3,
            "name": "b",
            "range": [
              2,
              3,
            ],
            "start": 2,
            "type": "Identifier",
          },
          "range": [
            0,
            4,
          ],
          "start": 0,
          "type": "MemberExpression",
        },
        "operator": "||",
        "range": [
          0,
          14,
        ],
        "right": {
          "end": 13,
          "left": {
            "computed": true,
            "end": 11,
            "object": {
              "end": 8,
              "name": "c",
              "range": [
                7,
                8,
              ],
              "start": 7,
              "type": "Identifier",
            },
            "optional": false,
            "property": {
              "end": 10,
              "name": "d",
              "range": [
                9,
                10,
              ],
              "start": 9,
              "type": "Identifier",
            },
            "range": [
              7,
              11,
            ],
            "start": 7,
            "type": "MemberExpression",
          },
          "operator": "=",
          "range": [
            7,
            13,
          ],
          "right": {
            "end": 13,
            "name": "e",
            "range": [
              12,
              13,
            ],
            "start": 12,
            "type": "Identifier",
          },
          "start": 7,
          "type": "AssignmentExpression",
        },
        "start": 0,
        "type": "LogicalExpression",
      },
      "range": [
        0,
        14,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 14,
  "range": [
    0,
    14,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > abc.package 1`] = `
{
  "body": [
    {
      "expression": {
        "computed": false,
        "object": {
          "name": "abc",
          "type": "Identifier",
        },
        "optional": false,
        "property": {
          "name": "package",
          "type": "Identifier",
        },
        "type": "MemberExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > abc.package 2`] = `
{
  "body": [
    {
      "expression": {
        "computed": false,
        "object": {
          "name": "abc",
          "type": "Identifier",
        },
        "optional": false,
        "property": {
          "name": "package",
          "type": "Identifier",
        },
        "type": "MemberExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > foo.bar 1`] = `
{
  "body": [
    {
      "expression": {
        "computed": false,
        "object": {
          "name": "foo",
          "type": "Identifier",
        },
        "optional": false,
        "property": {
          "name": "bar",
          "type": "Identifier",
        },
        "type": "MemberExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > x.y / z 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "computed": false,
          "object": {
            "name": "x",
            "type": "Identifier",
          },
          "optional": false,
          "property": {
            "name": "y",
            "type": "Identifier",
          },
          "type": "MemberExpression",
        },
        "operator": "/",
        "right": {
          "name": "z",
          "type": "Identifier",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Member > Expressions - Member (pass) > x[a, b] 1`] = `
{
  "body": [
    {
      "end": 7,
      "expression": {
        "computed": true,
        "end": 7,
        "object": {
          "end": 1,
          "name": "x",
          "range": [
            0,
            1,
          ],
          "start": 0,
          "type": "Identifier",
        },
        "optional": false,
        "property": {
          "end": 6,
          "expressions": [
            {
              "end": 3,
              "name": "a",
              "range": [
                2,
                3,
              ],
              "start": 2,
              "type": "Identifier",
            },
            {
              "end": 6,
              "name": "b",
              "range": [
                5,
                6,
              ],
              "start": 5,
              "type": "Identifier",
            },
          ],
          "range": [
            2,
            6,
          ],
          "start": 2,
          "type": "SequenceExpression",
        },
        "range": [
          0,
          7,
        ],
        "start": 0,
        "type": "MemberExpression",
      },
      "range": [
        0,
        7,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 7,
  "range": [
    0,
    7,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;
